atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java

```diff
@@ -456,8 +456,18 @@ private static boolean isConnectionException(Exception e) {
                 || e instanceof NoSuchElementException;
     }
 
+    private static boolean isOrContainsConnectionException(Exception e) {
+        if (e == null) {
+            return false;
+        } else if (isConnectionException(e)) {
+            return true;
+        } else {
+            return isOrContainsConnectionException((Exception) e.getCause());
+        }
+    }
+
     private static boolean isRetriableException(Exception e) {
```

Comment:
Thinking we might want to consider something along the lines of the following as `ClientCreationFailedException` is should be considered a connection exception:

```
    @VisibleForTesting
    static boolean isConnectionException(Throwable t) {
        return t != null
                && (t instanceof SocketTimeoutException
                        || t instanceof ClientCreationFailedException
                        || t instanceof UnavailableException
                        || t instanceof NoSuchElementException
                        || isConnectionException(t.getCause()));
    }

    @VisibleForTesting
    static boolean isRetriableException(Throwable t) {
        return t != null
                && (t instanceof TTransportException
                        || t instanceof TimedOutException
                        || t instanceof InsufficientConsistencyException
                        || isConnectionException(t)
                        || isRetriableException(t.getCause()));
    }
```